services/web/server/src/simcore_service_webserver/catalog_utils.py

```diff
@@ -0,0 +1,177 @@
+
+    # path & queries
+    backend_url = to_backend_service(
+        request.rel_url,
+        URL(settings.base_url),
+        settings.CATALOG_VTAG,
+    )
+    # FIXME: hack
+    if "/services" in backend_url.path:
+        backend_url = backend_url.update_query({"user_id": user_id})
+    logger.debug("Redirecting '%s' -> '%s'", request.url, backend_url)
+
+    # body
+    raw: Optional[bytes] = None
+    if request.can_read_body:
+        raw = await request.read()
+
+    # injects product discovered by middleware in headers
+    fwd_headers = request.headers.copy()
+    product_name = request[RQ_PRODUCT_KEY]
+    fwd_headers.update({X_PRODUCT_NAME_HEADER: product_name})
+
+    # forward request
+    return await catalog_client.make_request_and_envelope_response(
+        request.app, request.method, backend_url, fwd_headers, raw
+    )
+
+
+##  MODELS UTILS ---------------------------------
+
+
+def _get_unit_name(port: BaseServiceIOModel) -> str:
+    unit = port.unit
+    if port.property_type == "ref_contentSchema":
+        assert port.content_schema is not None  # nosec
+        unit = port.content_schema.get("x_unit", unit)
+        if unit:
+            # TODO: Review this convention under dev: x_units
+            # has a special format for prefix. tmp direct replace here
+            unit = unit.replace("-", "")
+        elif port.content_schema["type"] in ("object", "array"):
+            # these objects might have unit in its fields
+            raise NotImplementedError
+    return unit
+
+
+def _get_type_name(port: BaseServiceIOModel) -> str:
+    _type = port.property_type
+    if port.property_type == "ref_contentSchema":
+        assert port.content_schema is not None  # nosec
+        _type = port.content_schema["type"]
+    return _type
+
+
+@dataclass
+class UnitHtmlFormat:
+    short: str
+    long: str
+
+
+def get_html_formatted_unit(
+    port: BaseServiceIOModel, ureg: UnitRegistry
+) -> Optional[UnitHtmlFormat]:
+    try:
+        unit_name = _get_unit_name(port)
+        if unit_name is None:
+            return None
+
+        q = ureg.Quantity(unit_name)
+        return UnitHtmlFormat(short=f"{q.units:~H}", long=f"{q.units:H}")
+    except (PintError, NotImplementedError):
+        return None
+
+
+## PORT COMPATIBILITY ---------------------------------
+
+
+def _can_convert_units(from_unit: str, to_unit: str, ureg: UnitRegistry) -> bool:
+    assert from_unit  # nosec
+    assert to_unit  # nosec
+
+    # TODO: optimize by caching?  ureg already caches?
+    # TODO: symmetric
+    try:
+        return ureg.Quantity(from_unit).check(to_unit)
+    except (TypeError, PintError):
+        return False
+
+
+def can_connect(
+    from_output: ServiceOutput,
+    to_input: ServiceInput,
+    units_registry: UnitRegistry,
+) -> bool:
+    """Checks compatibility between ports.
+
+    This check IS PERMISSIVE and is used for checks in the UI where one needs to give some "flexibility" since:
```

Comment:
If I understand correctly this is to allow unit conversion of the same kind. Instead of `PERMISSIVE`  I would put it like this `PERMISSIVE(miles can connect to mm)`.